src/docquery/ext/pipeline.py

```diff
@@ -417,17 +415,31 @@ def postprocess_extractive_qa(
                 handle_impossible_answer,
                 max_answer_len,
             )
-
             word_ids = output["word_ids"]
             for s, e, score in zip(starts, ends, scores):
-                word_start, word_end = word_ids[s], word_ids[e]
-                if word_start is not None and word_end is not None:
+                if "token_logits" in output:
+                    predicted_token_classes = (
+                        output["token_logits"][
+                            0,
+                            s : e + 1,
+                        ]
+                        .argmax(axis=1)
+                        .cpu()
+                        .numpy()
+                    )
+                    assert np.setdiff1d(predicted_token_classes, [0, 1]).shape == (0,)
+                    token_indices = np.flatnonzero(predicted_token_classes) + s
+                else:
+                    token_indices = range(s, e + 1)
+
+                answer_word_ids = list(unique_everseen([word_ids[i] for i in token_indices]))
+                if len(answer_word_ids) > 0 and answer_word_ids[0] is not None and answer_word_ids[-1] is not None:
                     answers.append(
                         {
                             "score": float(score),
-                            "answer": " ".join(words[word_start : word_end + 1]),
-                            "start": word_start,
-                            "end": word_end,
+                            "answer": " ".join(words[i] for i in answer_word_ids),
+                            "words": answer_word_ids,
+                            # TODO: This is slightly inaccurate, since it is only the first page
```

Comment:
nit: Why is this inaccurate? Aren't these words all on a single page as per:
```
for page_idx, (image, word_boxes) in enumerate(input["pages"])
```
?

Reply:
Good point, removed. I was thinking about a different implementation.